src/core/thread/mesh_forwarder.cpp

```diff
@@ -2085,6 +2094,20 @@ void MeshForwarder::HandleFragment(uint8_t *aFrame, uint8_t aFrameLength,
             }
         }
 
+        // For a sleepy-end-device, if we receive a new (secure) next fragment
+        // with a non-matching fragmentation offset or tag, it indicates that
+        // we have either missed a fragment, or the parent has moved to a new
+        // message with a new tag. In either case, we can safely clear any
+        // remaining fragments stored in the reassembly list.
+
+        if (GetRxOnWhenIdle() == false)
+        {
+            if ((message == NULL) && (aMessageInfo.mLinkSecurity))
```

Comment:
The code reads pretty clear to me.  "if we receive a new (secure) next fragment with a non-matching fragmentation offset or tag".  Having `message == NULL` indicates we were unable to find a match.